dnsnameresolver.go

```diff
@@ -0,0 +1,289 @@
+							delete(resolver.wildcardDNSInfo, dnsName)
+						}
+					}
+				}
+				resolver.wildcardMapLock.Unlock()
+			} else {
+				// If the DNS name is regular, delete the details of the DNSNameResolver
+				// object from the regularDNSInfo map.
+				resolver.regularMapLock.Lock()
+				if dnsInfoMap, exists := resolver.regularDNSInfo[dnsName]; exists {
+					// If details of DNS name and the DNSNameResolver objects already exist
+					// then check if the existing information match with the current one.
+					// Otherwise, don't proceed.
+					if dnsInfoMap[resolverObj.Namespace] == resolverObj.Name {
+						delete(dnsInfoMap, resolverObj.Namespace)
+						if len(dnsInfoMap) > 0 {
+							resolver.regularDNSInfo[dnsName] = dnsInfoMap
+						} else {
+							delete(resolver.regularDNSInfo, dnsName)
+						}
+					}
+				}
+				resolver.regularMapLock.Unlock()
+			}
+
+			// Used only in unit tests.
+			if send != nil {
+				send(resolverObj)
+			}
+		},
+		// Used only in unit tests.
+		// Update event.
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// Get the DNSNameResolver object.
+			newResolverObj, ok := oldObj.(*ocpnetworkapiv1alpha1.DNSNameResolver)
+			if !ok {
+				log.Infof("object not of type DNSNameResolver: %v", oldObj)
+				return
+			}
+
+			// Check if namespace is configured or not.
+			if !resolver.configuredNamespace(newResolverObj.Namespace) {
+				return
+			}
+
+			// Used only in unit tests.
+			if send != nil {
+				send(newResolverObj)
+			}
+		},
+	})
+	return nil
+}
+
+// initPlugin initializes the ocp_dnsnameresolver plugin and returns the plugin startup and
+// shutdown callback functions.
+func (resolver *OCPDNSNameResolver) initPlugin() (func() error, func() error, error) {
+	// Create a client supporting network.openshift.io apis.
+	kubeConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	networkClient, err := ocpnetworkclient.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = resolver.initInformer(networkClient, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resolver.stopCh = make(chan struct{})
+
+	onStart := func() error {
+		go func() {
+			resolver.dnsNameResolverInformer.Run(resolver.stopCh)
+		}()
+
+		timeout := 5 * time.Second
+		timeoutTicker := time.NewTicker(timeout)
+		defer timeoutTicker.Stop()
+		logDelay := 500 * time.Millisecond
+		logTicker := time.NewTicker(logDelay)
+		defer logTicker.Stop()
+		checkSyncTicker := time.NewTicker(100 * time.Millisecond)
+		defer checkSyncTicker.Stop()
+		for {
+			select {
+			case <-checkSyncTicker.C:
+				if resolver.dnsNameResolverInformer.HasSynced() {
+					return nil
+				}
+			case <-logTicker.C:
+				log.Info("waiting for DNS Name Resolver Informer sync before starting server")
+			case <-timeoutTicker.C:
+				log.Warning("starting server with unsynced DNS Name Resolver Informer")
+				return nil
+			}
+		}
+	}
+
+	onShut := func() error {
+		resolver.stopLock.Lock()
```

Comment:
Same remark about the `stopLock`: why is it a struct field, it's only used inside `initPlugin`?